### PR TITLE
v1.10: mpi_f08: fix MPI_BUFFER_DETACH binding

### PR DESCRIPTION
**Pull request by jsquyres**
@ggouaillardet please review

(cherry picked from commit open-mpi/ompi@fc69c0be247857d7c5dc7d7c05bfaac2b8ae8589)

### CODE REVIEW

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/job/gh-ompi-release-v1.8-pr/368/

**Comment by rhc54** [comment]
@ggouaillardet could you please take a quick look at this? we need to release a 1.8.7 soon due to to the memory leak problem, and I'd like to sweep this into it

**Comment by ggouaillardet** [comment]
@jsquyres ompi_buffer_detach_f should be updated to get the correct result (see inline patch)
i am still wondering whether that should be f08 specific or not ...

BTW, should the C binding be 

```
MPI_Buffer_detach(void ** buffer, int * size)
```

instead of

```
MPI_Buffer_detach(void * buffer, int * size)
```

``` patch
diff --git a/ompi/mpi/fortran/mpif-h/buffer_detach_f.c b/ompi/mpi/fortran/mpif-h/buffer_detach_f.c
index bc88fd8..07010af 100644
--- a/ompi/mpi/fortran/mpif-h/buffer_detach_f.c
+++ b/ompi/mpi/fortran/mpif-h/buffer_detach_f.c
@@ -11,6 +11,8 @@
  *                         All rights reserved.
  * Copyright (c) 2007      Sun Microsystems, Inc.  All rights reserved.
  * Copyright (c) 2011-2012 Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2015      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
  * $COPYRIGHT$
  *
  * Additional copyrights may follow
@@ -67,17 +69,14 @@ OMPI_GENERATE_F77_BINDINGS (MPI_BUFFER_DETACH,

 void ompi_buffer_detach_f(char *buffer, MPI_Fint *size, MPI_Fint *ierr)
 {
-    /*
-     * It does not make sense in fortran to return a pointer
-     * here as the user may get a behavior that is unexpected.
-     * Therefore, we use a dummy variable and leave the value
-     * handed in alone.
-     */
     int c_ierr;
     void *dummy;
     OMPI_SINGLE_NAME_DECL(size);
     c_ierr = MPI_Buffer_detach(&dummy, OMPI_SINGLE_NAME_CONVERT(size));
-    if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
+    if (NULL != ierr) {
+        *(void **)buffer = dummy;
+        *ierr = OMPI_INT_2_FINT(c_ierr);
+    }

     if (MPI_SUCCESS == c_ierr) {
         OMPI_SINGLE_INT_2_FINT(size);
```

**Comment by jsquyres** [comment]
@ggouaillardet As for the C binding, check out the Advice to Users and Rationale on page 45 of MPI-3.1.  Short version: it's intentionally `(void*)` instead of `(void**)`.

Also per that ATU on page 45, the `buffer_addr` parameter is unused in the mpif.h and mpi module definition (!).  That is why `ompi_buffer_detach_f()` is the way it is (although it would be helpful for the comment in the current code to cite this).

So I guess we need an mpi_f08-specific C version of `ompi_buffer_detach_f()`...  Let me go code that up and update this PR.

Meh.  BUFFER_DETACH + Fortran = nightmare.

Good catch -- thanks!

**Comment by jsquyres** [comment]
@ggouaillardet Commit added to for the fixed mpi_f08 MPI_BUFFER_DETACH.  I also pushed some tests into the ibm test suite.

File: ompi/mpi/fortran/mpif-h/buffer_detach_f.c [hunk 1]
```diff
@@ -83,3 +86,25 @@ void ompi_buffer_detach_f(char *buffer, MPI_Fint *size, MPI_Fint *ierr)
...
+ */
+void ompi_buffer_detach_f08(char *buffer, MPI_Fint *size, MPI_Fint *ierr)
+{
+    int c_ierr;
+    void *dummy;
+
+    OMPI_SINGLE_NAME_DECL(size);
+    c_ierr = MPI_Buffer_detach(&dummy, OMPI_SINGLE_NAME_CONVERT(size));
+    if (NULL != ierr) {
+        *(void **)buffer = dummy;
```
Comment (ggouaillardet) [inline ompi/mpi/fortran/mpif-h/buffer_detach_f.c]: my bad...
this line should be moved to the next if block

Comment (jsquyres) [inline ompi/mpi/fortran/mpif-h/buffer_detach_f.c]: Oh!  Right!

**Comment by ggouaillardet** [comment]
:+1: 
btw, you did not update master yet, right ?
per @rhc54 a pr should me made for v1.8 too

**Comment by jsquyres** [comment]
Doing this for v1.8.x will be tricky -- there are ABI implications which I do not feel confident about addressing.  I think we should do this for v1.10 and forward, only.

I reset this PR (and the v2.x version -- #344) to reflect the 3 commits from master on which this work is based.

**Comment by jsquyres** [comment]
@ggouaillardet Since this went through a few revisions, can you give this PR another once-over?  And also the 2.x version (#344).  Thanks!

File: ompi/mpi/fortran/use-mpi-f08/pmpi-f-interfaces-bind.h [hunk 2]
```diff
@@ -89,8 +89,16 @@ subroutine pompi_buffer_attach_f(buffer,size,ierror) &
...
+! PMPI_BUFFER_DETACH (i.e., it is different than the mpif.h / mpi
+! module C implementation function).
+!
+! Note, too, we don't need a "p" version of the C implementation
+! function -- Fortran's interfaces provide MPI_ and PMPI_ names for
+! us; they can just both be bound to the same back-end
+! ompi_buffer_detach_f08 C function.
 subroutine pompi_buffer_detach_f(buffer_addr,size,ierror) &
-   BIND(C, name="pompi_buffer_detach_f")
+   BIND(C, name="ompi_buffer_detach_f08")
```
Comment (ggouaillardet) [inline ompi/mpi/fortran/use-mpi-f08/pmpi-f-interfaces-bind.h:101]: @jsquyres @bosilca 
should this be
subroutine pompi_buffer_detach_f08
instead?

Comment (ggouaillardet) [inline ompi/mpi/fortran/use-mpi-f08/pmpi-f-interfaces-bind.h:101]: @jsquyres @bosilca 
please disregard my previous comment
in ompi/mpi/fortran/mpif-h/profile/defines.h, a macro is missing for ompi_buffer_detach_f08

**Comment by ggouaillardet** [comment]
@jsquyres i guess the pragmas should be updated as well

``` c
#pragma weak PMPI_Buffer_detach_f08 = ompi_buffer_detach_f
#pragma weak MPI_Buffer_detach_f08 = ompi_buffer_detach_f
```

should be

``` c
#pragma weak PMPI_Buffer_detach_f08 = ompi_buffer_detach_f08
#pragma weak MPI_Buffer_detach_f08 = ompi_buffer_detach_f08
```

**Comment by ggouaillardet** [comment]
here is an other way to fix the issue reported by @bosilca 
(move ompi_buffer_detach_f08 function into the use-mpi-f08 directory)

``` patch
diff --git a/ompi/mpi/fortran/mpif-h/buffer_detach_f.c b/ompi/mpi/fortran/mpif-h/buffer_detach_f.c
index e2114d7..82b981b 100644
--- a/ompi/mpi/fortran/mpif-h/buffer_detach_f.c
+++ b/ompi/mpi/fortran/mpif-h/buffer_detach_f.c
@@ -29,7 +29,6 @@
 #pragma weak pmpi_buffer_detach__ = ompi_buffer_detach_f

 #pragma weak PMPI_Buffer_detach_f = ompi_buffer_detach_f
-#pragma weak PMPI_Buffer_detach_f08 = ompi_buffer_detach_f
 #elif OMPI_PROFILE_LAYER
 OMPI_GENERATE_F77_BINDINGS (PMPI_BUFFER_DETACH,
                            pmpi_buffer_detach,
@@ -47,7 +46,6 @@ OMPI_GENERATE_F77_BINDINGS (PMPI_BUFFER_DETACH,
 #pragma weak mpi_buffer_detach__ = ompi_buffer_detach_f

 #pragma weak MPI_Buffer_detach_f = ompi_buffer_detach_f
-#pragma weak MPI_Buffer_detach_f08 = ompi_buffer_detach_f
 #endif

 #if ! OPAL_HAVE_WEAK_SYMBOLS && ! OMPI_PROFILE_LAYER
@@ -72,7 +70,7 @@ OMPI_GENERATE_F77_BINDINGS (MPI_BUFFER_DETACH,
  * variable and leave the value handed in alone.
  *
  * The mpi_f08 implementation for MPI_BUFFER_DETACH is a separate
- * routine -- see below.
+ * routine.
  */
 void ompi_buffer_detach_f(char *buffer, MPI_Fint *size, MPI_Fint *ierr)
 {
@@ -87,23 +85,3 @@ void ompi_buffer_detach_f(char *buffer, MPI_Fint *size, MPI_Fint *ierr)
         OMPI_SINGLE_INT_2_FINT(size);
     }
 }
-
-/*
- * Per above, this is the mpi_f08 module implementation of
- * MPI_BUFFER_DETACH.  It handles the buffer arugment just like the C
- * binding.
- */
-void ompi_buffer_detach_f08(char *buffer, MPI_Fint *size, MPI_Fint *ierr)
-{
-    int c_ierr;
-    void *dummy;
-    OMPI_SINGLE_NAME_DECL(size);
-
-    c_ierr = MPI_Buffer_detach(&dummy, OMPI_SINGLE_NAME_CONVERT(size));
-    if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
-
-    if (MPI_SUCCESS == c_ierr) {
-        *(void **)buffer = dummy;
-        OMPI_SINGLE_INT_2_FINT(size);
-    }
-}
diff --git a/ompi/mpi/fortran/use-mpi-f08/Makefile.am b/ompi/mpi/fortran/use-mpi-f08/Makefile.am
index ec755f6..f382f2f 100644
--- a/ompi/mpi/fortran/use-mpi-f08/Makefile.am
+++ b/ompi/mpi/fortran/use-mpi-f08/Makefile.am
@@ -29,6 +29,9 @@ AM_FCFLAGS = -I$(top_builddir)/ompi/include \
              $(OMPI_FC_MODULE_FLAG). \
              -I$(top_srcdir) $(FCFLAGS_f90)

+AM_CPPFLAGS = -DOMPI_COMPILING_FORTRAN_WRAPPERS=1
+
+
 MOSTLYCLEANFILES = *.mod

 CLEANFILES += *.i90
@@ -803,7 +806,8 @@ libmpi_usempif08_la_SOURCES = \
         attr-fn-f08-callback-interfaces.h \
         conversion-fn-null-f08-interface.h \
         constants.h \
-        constants.c
+        constants.c \
+        ompi_buffer_detach_f08.c

 # These are generated; do not ship them
 nodist_libmpi_usempif08_la_SOURCES =
diff --git a/ompi/mpi/fortran/use-mpi-f08/ompi_buffer_detach_f08.c b/ompi/mpi/fortran/use-mpi-f08/ompi_buffer_detach_f08.c
new file mode 100644
index 0000000..8ca909a
--- /dev/null
+++ b/ompi/mpi/fortran/use-mpi-f08/ompi_buffer_detach_f08.c
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
+ *                         University Research and Technology
+ *                         Corporation.  All rights reserved.
+ * Copyright (c) 2004-2005 The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
+ *                         University of Stuttgart.  All rights reserved.
+ * Copyright (c) 2004-2005 The Regents of the University of California.
+ *                         All rights reserved.
+ * Copyright (c) 2007      Sun Microsystems, Inc.  All rights reserved.
+ * Copyright (c) 2011-2015 Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2015      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+#include "ompi/mpi/fortran/mpif-h/bindings.h"
+
+#if OPAL_HAVE_WEAK_SYMBOLS
+#pragma weak MPI_Buffer_detach_f08 = ompi_buffer_detach_f08
+#endif
+
+/*
+ * Per ompi/mpi/fortran/mpif-h/buffer_detach_f.c,
+ * this is the mpi_f08 module implementation of
+ * MPI_BUFFER_DETACH.  It handles the buffer arugment
+ * just like the C binding.
+ */
+void ompi_buffer_detach_f08(char *buffer, MPI_Fint *size, MPI_Fint *ierr)
+{
+    int c_ierr;
+    void *dummy;
+    OMPI_SINGLE_NAME_DECL(size);
+
+    c_ierr = MPI_Buffer_detach(&dummy, OMPI_SINGLE_NAME_CONVERT(size));
+    if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
+
+    if (MPI_SUCCESS == c_ierr) {
+        *(void **)buffer = dummy;
+        OMPI_SINGLE_INT_2_FINT(size);
+    }
+}
```

File: ompi/mpi/fortran/use-mpi-f08/pmpi-f-interfaces-bind.h (hunk 2 above)
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/pmpi-f-interfaces-bind.h:101]: Note the comment right above this code: it explains that we do not need a "p" version of ompi_buffer_detach_f08.  Specifically: this function is _only_ a back-end function; it is not exposed to users.  Hence, we don't need "MPI" and "PMPI" versions of it.  The Fortran bindings provide the name mangling from MPI and PMPI to the back-end ompi_buffer_detach_f08 for us.

**Comment by jsquyres** [comment]
@ggouaillardet Yeah, I think those f08 #pragmas should disappear from mpif-h/buffer_detach_f.c.

@ggouaillardet @bosilca As for which way to fix George's OS X issue, I think it's a po-TAY-to, po-TAH-to issue.  George is right that since we compile the ompi/mpi/fortran/mpif-h files twice, the f08 buffer detach function needs to be "turned off" for one of the compiler so that we don't get a duplicate symbol.  However, you can also make the argument that since it's an f08-specific function, it should be in the f08 dir (like Gilles' patch).  Since Gilles is working on revamping the Fortran bindings altogether, I don't think it matters much... Gilles, unless you have strong opinions, I'm going to go with George's fix of just adding the #if protection.

**Pull request merged**